### PR TITLE
Feat: merge ODK central work

### LINKED ISSUE #6235: Officially support ODK Central for offline stock movements
We released a tag [`v1.21.3-pcima`](https://github.com/IMA-WorldHealth/bhima/releases/tag/v1.21.3-pcima) with support for ODK Central.  This was to handle the PCIMA project's need to track stock offline with ODK forms.  I propose we formalize this support in BHIMA and merge the functionality into `master`.  It will make support easier going forwards.

We've released a module [`@ima-worldhealth/odk-central-api`](https://www.npmjs.com/package/@ima-worldhealth/odk-central-api) that should offload the ODK Central API functionality into its own tested module.  This means we can focus on rewriting the [`import.js`](https://github.com/IMA-WorldHealth/bhima/blob/v1.21.3-pcima/server/controllers/odk/import.js) functionality to bring stock movements into BHIMA from data collected in the ODK Central Server.

---

One thing to think about - we could migrate the pepkits project to using this version of BHIMA/Central and have another project using this functionality in addition to the PCIMA project.

### LINKED ISSUE #6302: Make the ODK QR code for a user available from BHIMA
We should make the QR code for configuring the ODK collect app available from within BHIMA.  If we plan on reusing the same users as we have in BHIMA, it would make sense to put them on the user's page.

### LINKED ISSUE #6322: Skip users without email addresses in odk sync
ODK sync crashes when a user doesn't have an email address.  We should skip users without email addresses.

### PR DESCRIPTION
**Pull request by jniles**
This PR adds in [ODK Central](https://docs.getodk.org/central-install/) functionality to the BHIMA server.  Here is what the ODK administration interface looks like:

image

You'll need to set your own username/password and ODK Central URL to test it out.

This installation is very bare-bones.  The way this works is somewhat documented in https://github.com/IMA-WorldHealth/bhima/issues/6235#issuecomment-1014303259.

This PR is plucked from work done by @jniles and @mbayopanda in `sprint-prosani`.

Closes #6235.
Closes #6302.
Closes #6322.

---

**TESTING**  (based on session with @jniles )
- Use bhima test (rebuild db, etc)
- Go to Administration > ODK Settings
  - Enter the ODK Central login credentials in the left-most panel (ODK Integration Settings)
  - In the middle panel, click on
    - Synchronize enterprise settings
    - Synchronize users
    - Synchronize forms
- Go to Adminstration > User Management
   - In the action menu for Superuser, click on the [ODK QR Code] link to display the QR code (leave it up on your screen)
- On your mobile phone
  - If you have not installed ODK Collect, do that
  - In ODK Collect, click on the [Configure with QR code] button on the entry screen. This should bring up the "Test Enterprise" page
  - Click on [Get Blank Form].  It should show that it has downloaded a form
  - On the bottom right of the page, click on the [Get Selected] button.  This downloads the form and sends you back to the "Test Enterprise" page.
  - Click on [Fill Blank Form] and select the form.  This should start filling out the form.
  - Fill out the form
     - For Nombre des articles, enter 1
     - For the barcode, first go to Bhima (Stock Lots) and display the barcode for lot VITAMIN-A.  Then scan that barcode with your phone.  When you click next, it should show the lot you selected.
     - Finish the form ([X] Mark form as finalize, press [Save Form and Exit])
  - Back on the "Test Enterprise" page, click [Send Finalized Form].  Select the form you just completed and press [Send Selected] on the bottom right of the page.
  - Back on Bhima go to Administration > ODK Settings
  - Click on:  Synchronize Submissions [Sync]
  - Go to Stock > Stock Movements and verify that you have a new stock entry in Depot Tertiaire

### CODE REVIEW

**Comment by jniles** [comment]
@jmcameron I requested a review too early.  There's some code I'm going to try and migrate out of this PR into the ODK Central API library.  I'll ping you when I'm ready for a review.

**Comment by jniles** [comment]
@jmcameron this is ready for review.

**Review by jmcameron — COMMENTED** [review]
Could not sync with the server.  Tried logging manually and that worked.  But when I carefully enter the ODK central server credentials and try "Sync Enterprise", I get 403 errors.

**Comment by jniles** [comment]
@jmcameron thanks for the review.  I think I gave you the wrong credentials.  I have sent you some new ones!

File: server/models/bhima.sql [hunk 1]
```diff
@@ -178,7 +178,8 @@ INSERT INTO unit VALUES
   (302, 'Cost Centers Accounts Report','TREE.COST_CENTER_ACCOUNTS_REPORT','Report of cc accounts values', 286,'/reports/cost_center_accounts'),
   (303, 'Cost Centers Balance Report','TREE.COST_CENTER_INCOME_EXPENSE_REPORT','Report of cc balance', 286,'/reports/cost_center_income_and_expense'),
   (304, '[SETTINGS] Settings', 'TREE.PAYROLL_SETTINGS', 'Payroll Settings', 57, '/payroll/setting'),
-  (305, 'Avg Medical Costs Per Patient', 'TREE.AVERAGE_MED_COST_REPORT', 'Report of avg med costs', 282, '/reports/avg_med_costs_per_patient');
+  (305, 'Avg Medical Costs Per Patient', 'TREE.AVERAGE_MED_COST_REPORT', 'Report of avg med costs', 282, '/reports/avg_med_costs_per_patient'),
+  (306, '[SETTINGS] ODK Settings', 'TREE.ODK_SETTINGS', 'ODK Settings', 1, '/admin/odk-settings');
```
Comment (jmcameron) [inline server/models/bhima.sql]: While using the [Settings] option to add the gear icon for the ODK Settings looks nice, it a bit odd in the context of the other Administration menu items.  Several of the items in that menu are for settings (eg Enterprise) and are not marked with the gear icon.  So the ODK Settings is not consistent with the rest.  So I suggest either removing [SETTINGS] or adding to other items in the Administration menu that are for settings.  Also, in reality, the ODK Settings item is about more than settings.  Only the left most panel is for settings.  The rest are for actions, so this page is not ALL settings.

File: server/models/migrations/next/migrate.sql [hunk 2]
```diff
@@ -1,11 +1,35 @@
-/**
- * @author: lomamech
- * @description: update migrate.sql file for
- * @date: 2022-01-24
-*/
-UPDATE report SET `report_key` = 'analysis_auxiliary_cashbox', `title_key` = 'REPORT.ANALYSIS_AUX_CASHBOX.TITLE'
```
Comment (jmcameron) [inline server/models/migrations/next/migrate.sql:6]: It was not obvious why these got deleted.

**Review by jmcameron — COMMENTED** [review]
I found the term "ODK Central" a little vague.  As I was trying to figure out ODK, it was not obvious if this was part of BHIMA or not.  So I suggest renaming this to "BHIMA ODK Server" or something similar.

Obviously this is a bare-bones ODK server.   However (in another PR), we need a registry for ODK forms, submissions, aggregated data, etc.

I did a basic review of the code.  See the inline code comments.

I also tested this (per the procedure outlined in the PR description) with assistance from @jniles.  Other than an online server database issue that @jniles fixed behind the scenes, it all worked nicely.

We found that the QR code for the ODK server settings works well.  However, the BHIMA lot barcodes seem to be a little small for reliable scanning on my cell phone.  So I suggest increasing the lot barcode display.  If that is not done in this PR, I suggest creating a new issue for it.

In practice, I suspect the layout of this page will evolve.  It seems like it should perhaps be a "ODK Campaign" registry and the Settings and sync actions might fall under the main Menu button?

File: server/models/bhima.sql (hunk 1 above)
Comment (jniles) [reply server/models/bhima.sql]: Yeah, this is a good point.  Let me do that.

**Comment by jniles** [comment]
@jmcameron thanks for the review.  ODK Central is the official name: https://docs.getodk.org/central-setup/.  I do agree that it is confusing.  They used to have a different server called ODK Aggregate, which also was a pretty non-googleable name.

File: server/models/migrations/next/migrate.sql (hunk 2 above)
Comment (jniles) [reply server/models/migrations/next/migrate.sql:6]: Probably a failed merge.  Let me fix it.

**Comment by jmcameron** [comment]
> @jmcameron thanks for the review. ODK Central is the official name: https://docs.getodk.org/central-setup/. I do agree that it is confusing. They used to have a different server called ODK Aggregate, which also was a pretty non-googleable name.

Got it.  Somehow I thought "ODK Central" was a server that we actually set up ourselves.

**Comment by jniles** [comment]
@jmcameron got the assignment bug fixed.  If you want to test it again, feel free.  Otherwise, I think this is ready to go.

**Comment by jmcameron** [comment]
Bors r+

**Comment by jmcameron** [comment]
@jniles It was not obvious why the merge attempt failed.  I'll try again.

**Comment by jniles** [comment]
Thanks!

**Pull request merged**